bool has_x
uint32 x
bool has_y
uint32 y
